trinodb/trino
# Multi-join pushdown appends recursively _<number> to column names so that they exceed maximum alias length

@wendigo, @hashhar 
# How to Reproduce

- Start ‘OracleQueryRunner’ in debug mode.
- Setup a breakpoint in DefaultJdbcMetadata.applyJoin() method where the code is calling jdbcClient.implementJoin().
- Start ‘Trino’ console client.
```
trino> use oracle.trino_test;
```
The following enables join pushdown that otherwise would not happen [1], [2].
```
trino> set session oracle.join_pushdown_enabled=true;
```
The following query
```
trino> SELECT c.custkey, o.orderkey, n.nationkey
      FROM orders o JOIN customer c ON c.custkey = o.custkey
      JOIN nation n ON c.nationkey = n.nationkey;
```
should trigger the breakpoint.

<img width="1303" alt="Screenshot 2023-08-09 at 15 28 12" src="https://github.com/trinodb/trino/assets/13519572/d1705623-941b-4661-90eb-5bc3e4f9008d">

In ‘Threads & Variables’ view you should be able to observe the ‘… AS COLUMN_X_Y’ aliases for columns that have as many underscores with a follow up number, as the number of joins in the query.

# Proposed solution
Divide the number of characters available for the alias (N) into two:
- ‘Column Name Prefix’ of length P
- ‘Uniqueness Magic Number’ of length M

where N - 1 = P + M. Use the additional character '_' to visually divide 'Original Column Name Prefix' from 'Uniqueness Magic Number'.

__Example__

Assume N = 8, P = 4, M = 3

Query: 

```
SELECT c.custkey, o.orderkey
  FROM orders o JOIN customer c
  ON c.custkey = o.custkey
```
The aliases would then become:
- custkey => cust_001
- orderkey => orde_002
